File: 백준/Silver/9012. 괄호/괄호.java
import java.util.*;
import java.io.*;

public class Main {
    static int n;
    static LinkedList<Character> st;

    public static void main(String[] args) throws Exception {
        //System.setIn(new FileInputStream("res/boj.txt"));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        StringBuilder sb = new StringBuilder();
        n = Integer.parseInt(st.nextToken());
        for (int i = 0; i < n; i++) {
            if (go(br.readLine())) sb.append("YES").append("\n");
            else sb.append("NO").append("\n");
        }
        System.out.println(sb);
    }

    public static Boolean go(String input) {
        st = new LinkedList<>();
        for (int i = 0; i < input.length(); i++) {
            // ( 랑 ) 가 들어오는데, 닫는게 들어오는 거면 넣지않고 pop, 여는게 들어오는 그냥 push
            if (input.charAt(i) == ')') {
                if (st.isEmpty()) return false;
                st.pop();
            } else {
                st.push(input.charAt(i));
            }
        }

        if (st.isEmpty()) return true;

        return false;
    }
}

File: 백준/Silver/9012. 괄호/괄호.kt
fun main()=with(System.`in`.bufferedReader()){
    val size = readLine().toInt()
    val writer = System.out.bufferedWriter()
    for(i in 0 until size){
        val input = readLine().toCharArray()
        var closed = 0 // ( = ++, ) = --
        for(j in input){
            if(j == '(') closed++
            else{
                closed--
                if(closed<0){
                    break
                }
            }
        }
        if(closed==0) writer.write("YES\n") else writer.write("NO\n")
    }

    writer.flush()
    writer.close()
    close()
}